File: frontend/src/api/production.ts
import type {
  BatchPublic,
  FormulaCreate,
  FormulaLinePublic,
} from "../types/api";
import { http } from "./client";

export async function getFormula(
  productId: number
): Promise<FormulaLinePublic[]> {
  const { data } = await http.get<FormulaLinePublic[]>(
    `/production/formula/${productId}`
  );
  return data;
}

export async function saveFormula(
  payload: FormulaCreate
): Promise<FormulaLinePublic[]> {
  const { data } = await http.put<FormulaLinePublic[]>(
    "/production/formula",
    payload
  );
  return data;
}

export async function listBatches(
  productId?: number
): Promise<BatchPublic[]> {
  const { data } = await http.get<BatchPublic[]>("/production/batches", {
    params: productId !== undefined ? { product_id: productId } : {},
  });
  return data;
}

export async function registerBatch(
  payload: Omit<BatchPublic, "id" | "produced_at" | "created_by_id">
): Promise<BatchPublic> {
  const { data } = await http.post<BatchPublic>(
    "/production/batches",
    payload
  );
  return data;
}

File: backend/app/api/routes/production.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.routes.auth import get_current_user
from app.core.database import get_db
from app.models.user import User
from app.schemas.production import (
    BatchCreate,
    BatchPublic,
    FormulaCreate,
    FormulaLinePublic,
)
from app.services.production_service import ProductionService

router = APIRouter(
    prefix="/production",
    tags=["production"],
    dependencies=[Depends(get_current_user)],
)


@router.get(
    "/formula/{product_id}", response_model=list[FormulaLinePublic]
)
def get_formula(product_id: int, db: Session = Depends(get_db)):
    return ProductionService(db).get_formula(product_id)


@router.put(
    "/formula", response_model=list[FormulaLinePublic], status_code=200
)
def save_formula(payload: FormulaCreate, db: Session = Depends(get_db)):
    return ProductionService(db).save_formula(payload)


@router.get("/batches", response_model=list[BatchPublic])
def list_batches(
    product_id: int | None = None, db: Session = Depends(get_db)
):
    return ProductionService(db).list_batches(product_id)


@router.post("/batches", response_model=BatchPublic, status_code=201)
def register_batch(
    payload: BatchCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return ProductionService(db).register_batch(payload, current_user.id)
